DDS-1218-reporter

dmlond opened the pull request:
Reporter code first past. Specs need more coverage.

dmlond committed:
merged upstream

added functions (still need tests)
- download_url
- file_version
- upload
- chunk_text
- upload_md5
- report_md5

dmlond committed:
dds_md5_reporter is functional, submitting for WIP Code Review
dds_md5_reporter_spec has some coverage, but needs much more coverage

dmlond committed:
previous commit linked to an uncommitted file

dmlond committed:
factored dds_api HTTParty calls into method
needs testing

dmlond committed:
further refinement of dds_api usage

dmlond committed:
all HTTParty calls factored into call_external method
needs tests

dmlond committed:
refactored to mock dds_api and call_external calls in shared_examples for failed calls

dmlond committed:
add test for call_external

dmlond committed:
add test for dds_api

dmann commented on spec/dds_md5_reporter_spec.rb:
```diff
@@ -6,27 +6,59 @@
+      it 'should raise an error' do
+        expect {
+          subject
+        }.to raise_error ArgumentError, "missing file_version_id, file_version_id, user_key, agent_key, and dds_api_url cannot be nil"
```
I find this error message confusing to read, mostly because of the first comma. Maybe something like `missing file_version_id. The keywords file_version_id, user_key, agent_key, and dds_api_url cannot be nil.`

dmlond replied:
ok

dmlond committed:
changed to reflect PR code review suggestion

dmann commented on spec/dds_md5_reporter_spec.rb:
```diff
@@ -90,14 +122,66 @@
+            }
+          }
+          let(:expected_body) {
+            dds_error_report.to_json
```
Does `to_json` create a single or multi-line json string? I *think* that DDS serializes to multi-line json, but I'm not sure about that.

dmlond replied:
I built the expectation off of actual testing. The main expectation is that:
- resp.body.match(/error.*reason.*suggestion/) returns true
- resp.parsed_response deserializes the body to a hash with stringified keys
I verified both on an actual responses from calling the dds api

dmann replied:
Ok, just wanted to be sure, since ruby regex is picky about single vs multi-line strings.

dmlond committed:
modified initializer error based on code review comments

add shared_example 'a method with a cached response' which
tests that the method only calls dds_api once

add tests for file_version, download_url, and upload

dmann commented on spec/dds_md5_reporter_spec.rb:
```diff
@@ -164,42 +227,291 @@
+
+            it {
+              travel_to(Time.now + 10000) do
+                expect(subject.auth_token).to eq(expected_api_token)
```
I'm finding the `#auth_token` tests difficult to follow, this one in particular. Please add descriptions to the `it` blocks.

dmlond replied:
I will refactor this test. I really like using travel_to, but it isnt really apparent that the test expects the auth_token to be refreshed with a new call to the dds api when the token is expired.

dmlond committed:
add test for chunk_text

dmann commented on spec/dds_md5_reporter_spec.rb:
```diff
@@ -164,42 +227,291 @@
+        expect(reporter).to receive(:raise_dds_api_exception)
+          .with(expected_preamble, expected_response)
+          .and_raise(expected_error)
+        if (expected_http_verb == :get)
```
It's best to avoid branching logic in tests.

dmlond replied:
ok. I removed this already, as I refactored this another way.

dmann commented on spec/dds_md5_reporter_spec.rb:
```diff
@@ -164,42 +227,291 @@
+          .with(expected_preamble, expected_response)
+          .and_raise(expected_error)
+        if (expected_http_verb == :get)
+          expect(HTTParty).to receive(expected_http_verb)
```
It would be great to find a way to test without "mocking what you don't own".

dmlond replied:
in the yet to be pushed version I wrap all HTTParty calls in a single method, and I only mock the HTTParty methods in the tests for that method :)

dmlond committed:
move shared_examples and shared_contexts to beginning for reuse in #auth_token
refactor #auth_token tests:
  - use a failed dds_api response shared behavior for incorrect credentials
  - use a method with a cached response shared behavior for subsequent call before
    cached token has expired
  - create better documented test for when subsequent call is made
    after cached token has expired

dmlond committed:
refactor a failed external call (and a failed dds api call) shared examples
and tests that use them to remove branching logic from the test

dmlond committed:
refactored a method with a cached response behavior to use call_external_arguments
instead of expected_dds_api_arguments

refactored tests that behave like a method with a cached response to
remove expected_dds_api_arguments since call_external_arguments were
defined more globally to the test already

refactored #auth_token test for expired to token to also use
call_external_arguments

dmlond committed:
modify documentation grammer error for #auth_token test when token has expired

refactor tests to remove let params that are no longer needed after refactoring
to use shared call_external_arguments

refactor download_url test to be more consistent with other tests
that make external calls

dmlond committed:
further DRYing by creating successful call behaviors and contexts with parsed_response and body expectations
use successful call behaviors or contexts in all tests instead of repeating expectations

dmlond committed:
add tests for chunk_text and report_md5. This completes test coverage

dmlond committed:
factoring response success check into its own method for easier testing and mocking

dmann reviewed (approved):
:shipit:
